dftbplus: Update and add upstream maintainer

HaoZeke opened the pull request:
As noted. Part of the CECAM ESL initiative of upstreaming recipes for foundational packages.

HaoZeke committed:
dftbp: Update and add upstream maintainer

HaoZeke committed:
dftbp: Trust in the hybrid cmake builds

HaoZeke committed:
dftbp: Handle scalapack better

HaoZeke commented:
FWIW the test failures seem to be unrelated (`pod not found`)

tldahlgren commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,160 +3,164 @@
-        "gpu",
-        default=False,
-        description="Use the MAGMA library " "for GPU accelerated computation",
+        "build_type",
```
FYI. Usually `variant` directives appear before `depends_on` directives to show the build options being exposed to the install process before defining dependencies, which may be depend on those variants.

I'm curious as to the motivation for putting the dependencies first.

HaoZeke replied:
Ah, my bad, I didn't notice the convention but that makes sense. The motivation as such was just that the `variant` directives were visually more noisy. Updated.

tldahlgren reviewed (commented):
Confirmed all sha256.

tldahlgren commented:
@aradi Are you willing to be a maintainer on this package?

HaoZeke committed:
dftbp: Refactor as per review

Co-authored-by: Tamara Dahlgren <dahlgren1@llnl.gov>

tldahlgren commented:
I can build the newer version successfully but the switch to `CMakePackage` does not work for version `19.1`:

```
...
==> Installing dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl
==> No binary for dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/4d/4d07f5c6102f06999d8cfdb1d17f5b59f9f2b804697f14b3bc562e3ea094b8a8.tar.gz
==> No patches needed for dftbplus
==> dftbplus: Executing phase: 'cmake'
==> Error: ProcessError: Command exited with status 1:
    '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/cmake-3.23.2-ogokiotf7a2l42brsywrxk4hagfryv3d/bin/cmake' '-G' 'Ninja' '-DCMAKE_INSTALL_PREFIX:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl/lib64;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/openblas-0.3.19-z24tobvjxwysziul3oc3kujz6tcgib45/lib' '-DCMAKE_PREFIX_PATH:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/openblas-0.3.19-z24tobvjxwysziul3oc3kujz6tcgib45;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/ninja-1.10.2-znkui4c4n7dl7ml5h5n2w3734qyjy4xh;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/cmake-3.23.2-ogokiotf7a2l42brsywrxk4hagfryv3d' '-DWITH_API:BOOL=ON' '-DWITH_ARPACK:BOOL=OFF' '-DWITH_CHIMES:BOOL=OFF' '-DWITH_ELSI:BOOL=OFF' '-DWITH_GPU:BOOL=OFF' '-DWITH_MBD:BOOL=OFF' '-DWITH_MPI:BOOL=OFF' '-DWITH_PLUMED:BOOL=OFF' '-DWITH_POISSON:BOOL=OFF' '-DWITH_PYTHON:BOOL=OFF' '-DWITH_SDFTD3:BOOL=OFF' '-DWITH_SOCKETS:BOOL=OFF' '-DWITH_TBLITE:BOOL=OFF' '-DWITH_TRANSPORT:BOOL=OFF' '-DWITH_UNIT_TESTS=OFF' '/var/tmp/dahlgren/spack-stage/spack-stage-dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl/spack-src'

1 error found in build log:
     3    CMake Warning:
     4      Ignoring extra path from command line:
     5    
     6       "/var/tmp/dahlgren/spack-stage/spack-stage-dftbplus-19.1-epzt6sm3z
          g3ne3kg3w33d4b6ac42e5gl/spack-src"
     7    
     8    
  >> 9    CMake Error: The source directory "/var/tmp/dahlgren/spack-stage/spac
          k-stage-dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl/spack-src" doe
          s not appear to contain CMakeLists.txt.
     10   Specify --help for usage, or press the help button on the CMake GUI
...
```

HaoZeke commented:
> I can build the newer version successfully but the switch to `CMakePackage` does not work for version `19.1`:
> 
> [Block (23 lines)](#)
> ```
> ...
> ==> Installing dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl
> ==> No binary for dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl found: installing from source
> ==> Fetching https://mirror.spack.io/_source-cache/archive/4d/4d07f5c6102f06999d8cfdb1d17f5b59f9f2b804697f14b3bc562e3ea094b8a8.tar.gz
> ==> No patches needed for dftbplus
> ==> dftbplus: Executing phase: 'cmake'
> ==> Error: ProcessError: Command exited with status 1:
>     '/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/cmake-3.23.2-ogokiotf7a2l42brsywrxk4hagfryv3d/bin/cmake' '-G' 'Ninja' '-DCMAKE_INSTALL_PREFIX:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl/lib;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl/lib64;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/openblas-0.3.19-z24tobvjxwysziul3oc3kujz6tcgib45/lib' '-DCMAKE_PREFIX_PATH:STRING=/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/openblas-0.3.19-z24tobvjxwysziul3oc3kujz6tcgib45;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/ninja-1.10.2-znkui4c4n7dl7ml5h5n2w3734qyjy4xh;/usr/WS1/dahlgren/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/cmake-3.23.2-ogokiotf7a2l42brsywrxk4hagfryv3d' '-DWITH_API:BOOL=ON' '-DWITH_ARPACK:BOOL=OFF' '-DWITH_CHIMES:BOOL=OFF' '-DWITH_ELSI:BOOL=OFF' '-DWITH_GPU:BOOL=OFF' '-DWITH_MBD:BOOL=OFF' '-DWITH_MPI:BOOL=OFF' '-DWITH_PLUMED:BOOL=OFF' '-DWITH_POISSON:BOOL=OFF' '-DWITH_PYTHON:BOOL=OFF' '-DWITH_SDFTD3:BOOL=OFF' '-DWITH_SOCKETS:BOOL=OFF' '-DWITH_TBLITE:BOOL=OFF' '-DWITH_TRANSPORT:BOOL=OFF' '-DWITH_UNIT_TESTS=OFF' '/var/tmp/dahlgren/spack-stage/spack-stage-dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl/spack-src'
> 
> 1 error found in build log:
>      3    CMake Warning:
>      4      Ignoring extra path from command line:
>      5    
>      6       "/var/tmp/dahlgren/spack-stage/spack-stage-dftbplus-19.1-epzt6sm3z
>           g3ne3kg3w33d4b6ac42e5gl/spack-src"
>      7    
>      8    
>   >> 9    CMake Error: The source directory "/var/tmp/dahlgren/spack-stage/spac
>           k-stage-dftbplus-19.1-epzt6sm3zg3ne3kg3w33d4b6ac42e5gl/spack-src" doe
>           s not appear to contain CMakeLists.txt.
>      10   Specify --help for usage, or press the help button on the CMake GUI
> ...
> ```

Yikes, that's a mess. Is there a best practice for this situation? Inherit from both `MakefilePackage` and `CMakePackage` and use a `conflicts()` or something?

HaoZeke committed:
dftbp: (man)handle version 19.1

aradi commented:
@tldahlgren Yes, I would be happy to co-maintain this package.

HaoZeke committed:
dftbp: Build shared for python

HaoZeke commented:
@tldahlgren any updates on this?

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
         "(serial version is built without libNEGF/transport)",
     )
+    variant(
+        "tblite", default=False, description="Whether xTB support should be included via tblite."
```
There is no tblite in spack, so DFTB+ will fetch the version from GH when building with this variant enabled. Also, support for this has been added in 21.2 (add `when="@21.2:"`).

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+        description="Whether the Python components of DFTB+ should be tested and installed.",
+    )
+    variant(
+        "sdftd3",
```
Has been added in version 21.2 (add `when="@21.2:"`)

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+
+    # ONLY for version 19.1
     variant(
         "dftd3",
```
Has been deprecated in version 21.2 (add a `when="@:21.1"`)

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+    )
+    variant("mpi", default=False, description="Whether DFTB+ should support MPI-parallelism.")
+    variant(
+        "omp", default=True, description="Whether OpenMP thread parallisation should be enabled."
```
Most packages seem to use `openmp`

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+        description="Use the MAGMA library " "for GPU accelerated computation",
+    )
+    variant(
+        "mbd",
```
The required dependency libmbd is not in spack, DFTB+ will fetch it when building. Also, this has been added in version 21.1 (add `when="@21.1:"`)

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+        description="Whether the ARPACK library should be included (needed for TD-DFTB).",
+    )
+    variant(
+        "chimes",
```
Support for chimes has been added in 21.2 (add `when="@21.2:"`)

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
-        sha256="bd191b3d240c1a81a8754a365e53a78b581fc92eb074dd5beb8b56a669a8d3d1",
-        destination="external/slakos",
-        when="@18.2:",
+    url = "https://github.com/dftbplus/dftbplus/archive/22.1.tar.gz"
```
That's the wrong URL, there is a dedicated release artifact which is not autogenerated via git-archive.

https://github.com/dftbplus/dftbplus/releases/download/22.2/dftbplus-22.2.tar.xz

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+        "omp", default=True, description="Whether OpenMP thread parallisation should be enabled."
+    )
+    variant(
+        "plumed",
```
Support added in version 20.1 (add `when="@20.1:"`)

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+    )
+    variant("poisson", default=False, description="Whether the Poisson-solver should be included.")
+    variant(
+        "python",
```
Enabling the Python variant currently does not add the installed Python files to the `PYTHONPATH`, therefore `dftbplus` won't be importable. Add ``extends("python", when="+python")``.

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+        default=False,
+        description="Whether metadynamics via the PLUMED2 library should be allowed.",
+    )
+    variant("poisson", default=False, description="Whether the Poisson-solver should be included.")
```
Should the `-poisson` variant conflict with `+transport`.

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
+        "tblite", default=False, description="Whether xTB support should be included via tblite."
+    )
 
+    variant("shared", default=False, description="Most often for the Python wrappers.")
```
The variant `-shared` needs to conflict with `+python` based on the description.

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
-    depends_on("scalapack", when="+mpi")
-    depends_on("mpi", when="+mpi")
+    depends_on("cmake@3.16:", type="build")
+    depends_on("ninja", type="build")
```
Should be `ninja@1.10:` otherwise CMake cannot use ninja for building.

awvwgk commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,142 @@
 
-    variant("mpi", default=True, description="Build an MPI-paralelised version of the code.")
+    variant(
+        "build_type",
```
Why is this needed? Isn't the `build_type` variant inherited from the `CMakePackage` base class?

awvwgk reviewed (changes requested):
Just noticed some issues with the variants specified here.

zackgalbreath commented:
@spackbot run pipeline

fangohr commented:
This merge request was never completed. There is a similar one (and more recent one) at https://github.com/spack/spack/pull/39608 .

@iamashwin99, @HaoZeke, @aradi - could you join forces to get one of them through? We need DFTBPLUS and the features in https://github.com/spack/spack/pull/39608  for Octopus (and will send a merge request for an update of the Octopus package once either of this PRs has come through).

If there is no interest to re-activate this somewhat stale PR, perhaps @iamashwin99 should change https://github.com/spack/spack/pull/39608 from draft and the review should focus on that?

HaoZeke commented:
It wasn't wholly clear what the hold up on this PR was. We got the package creator's support as well... Then it seems to have just stalled?

fangohr commented:
I guess there were unanswered queries, and that's why it may have stalled?  By now, the logs cannot be seen anymore so it is hard to debug this without running it again. 

@iamashwin99 - could you integrate your changes from https://github.com/spack/spack/pull/39608 in this PR, or would you prefer to ask for https://github.com/spack/spack/pull/39608 to be considered for merging? (I am not sure how similar the 2 MRs are.)

awvwgk commented:
> It wasn't wholly clear what the hold up on this PR was. We got the package creator's support as well... Then it seems to have just stalled? 

There are a couple of unaddressed comments from the review, especially those regarding incorrect dependency versions should be addressed.

HaoZeke committed:
dftbp: Address review comments

Co-authored-by: awvwgk <awvwgk@users.noreply.github.com>

HaoZeke committed:
dftbp: Add another maintainer

Co-authored-by: iamashwin99 <iamashwin99@users.noreply.github.com>

HaoZeke commented:
> > It wasn't wholly clear what the hold up on this PR was. We got the package creator's support as well... Then it seems to have just stalled?
> 
> There are a couple of unaddressed comments from the review, especially those regarding incorrect dependency versions should be addressed.

Ah, sorry my bad, I think I got all of them now, and I added @iamashwin99 as a co-maintainer, but since from what I could see in a quick look this PR seems to be more comprehensive (variants, conflicts, maintainers) than the other.

HaoZeke committed:
dftp: Fix typo

iamashwin99 commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,158 @@
+    depends_on("blas", when="-mpi")
+    depends_on("lapack", when="-mpi")
+
+    depends_on("arpack-ng", when="+arpack")
```
https://github.com/dftbplus/dftbplus/blob/97c45a0eea394cc2fd97538e4c5ae6052e4fcb23/config.cmake#L37
Works only with non-MPI (serial) build, needed for Casida linear response
```suggestion
    depends_on("arpack-ng", when="+arpack~mpi")
```

iamashwin99 commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,158 @@
+        "arpack",
         default=False,
-        description="Use the MAGMA library " "for GPU accelerated computation",
+        description="Whether the ARPACK library should be included (needed for TD-DFTB).",
+    )
```
Works only with non-MPI (serial) build, needed for Casida linear response
https://github.com/dftbplus/dftbplus/blob/97c45a0eea394cc2fd97538e4c5ae6052e4fcb23/config.cmake#L37
```suggestion
        description="Whether the ARPACK library should be included (needed for TD-DFTB).",
        when="~mpi"
    )
```

awvwgk replied:
Should that be
```suggestion
        description="Whether the ARPACK library should be included (needed for TD-DFTB).",
        when="~mpi",
    )
```

iamashwin99 replied:
you are right! I have corrected the typo

iamashwin99 commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,158 @@
     variant(
         "elsi",
         default=False,
         description="Use the ELSI library for large scale systems. "
         "Only has any effect if you build with '+mpi'",
     )
```
Only has any effect if you build with '+mpi', so we can make the variant also available only when mpi build is built
```suggestion
        description="Use the ELSI library for large scale systems. "
        "Only has any effect if you build with '+mpi'",
        when="+mpi"
    )
```

iamashwin99 commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,158 @@
-
-    variant("arpack", default=False, description="Use ARPACK for excited state DFTB functionality")
-
     variant(
         "transport",
         default=False,
         description="Whether transport via libNEGF should be included. "
         "Only affects parallel build. "
         "(serial version is built without libNEGF/transport)",
     )
```
This should be build only with "+shared" variant.
see https://github.com/dftbplus/dftbplus/blob/97c45a0eea394cc2fd97538e4c5ae6052e4fcb23/config.cmake#L25

aradi replied:
But wait, isn't `+shared` exactly the opposite, what the documentation requires (namely the you build a **static** library)?

But anyway, this seems to be a leftover in the docs from our side, as we use libNEGF via Conda and it is linked there as a shared library as you can see in the [Conda-recipe](https://github.com/conda-forge/dftbplus-feedstock/blob/main/recipe/build.sh). (Sorry for the confusion, I'll remove that comment from our `config.cmake`)

iamashwin99 commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,158 @@
+    depends_on("mpi", when="+mpi")
+    depends_on("plumed", when="+plumed")
+    depends_on("scalapack", when="+mpi")
+    depends_on("python", when="+python")
```
We can make this a little more concrete with specifications from 
- python>3.2 https://github.com/dftbplus/dftbplus/blob/e724a2d5f90bed716337abafbf9cc2884e535cca/INSTALL.rst#L105
- numpy requirement: https://github.com/dftbplus/dftbplus/blob/e724a2d5f90bed716337abafbf9cc2884e535cca/tools/dptools/README#L65
```suggestion
    depends_on("python@3.2:", type=("build", "run"))
    depends_on("py-numpy", type=("build", "run"))  # for tests
```

iamashwin99 commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -160,3 +233,56 @@ def edit(self, spec, prefix):
+            self.define_from_variant("WITH_TRANSPORT", "transport"),
+            self.define_from_variant("BUILD_SHARED_LIBS", "shared"),
+        ]
+        # SCALAPACK
```
Can we add this comment from my MR, maybe this adds a bit more information to the future maintainers.
```suggestion
        # SCALAPACK
        # Note: dftbplus@20.1 uses plural form of the option names
        #       (e.g. -DSCALAPACK_LIBRARIES)
        # but dftbplus@20.2 onwards uses singular
        #       (e.g. -DSCALAPACK_LIBRARY)
        # and plural form is ignored.
        # We set both inorder to be compatible with all versions.
```

aradi replied:
Sorry for that, only after we have changed to CMake and made the first release, did I realize, that projects are not supposed to use the plural form, which are by convention "reserved" for CMake itself. (Of course, this info is well hidden in the CMake docs...) Do we need to support old version at all? Why not starting with 20.2 or even 21.1? Since then, our CMake interface is quite stable. (I don't see any reason, why somebody wanted to install a DFTB+ version which is more than 2 years old.)

iamashwin99 replied:
The ESL bundle still ships with dftbplus@20.1 :sweat_smile:   https://esl.cecam.org/en/bundle/

aradi replied:
OK, well spotted. :smile: But probably, that should not be the reference (or are there any plans to create a spack recipe for the ESL-bundle?)

iamashwin99 reviewed (changes requested):
thanks a lot for the changes!
this MR is more comprehensive with a lot of variants.
I have not tested the compilation of this branch yet, I will do so and then update.

There are few more details that I suggested based on my experience with the other MR.
dftb+ has a lot of dependencies which also changes over versions and its tricky to keep track of them :)

iamashwin99 commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -160,3 +233,56 @@ def edit(self, spec, prefix):
+            self.define_from_variant("BUILD_SHARED_LIBS", "shared"),
+        ]
+        # SCALAPACK
+        if "+mpi" in self.spec:
+            args.extend(
+                [
+                    self.define("SCALAPACK_FOUND", "true"),
+                    self.define("SCALAPACK_INCLUDE_DIRS", self.spec["scalapack"].prefix.include),
+                    self.define("SCALAPACK_LIBRARIES", self.spec["scalapack"].libs.joined(";")),
+                    self.define("SCALAPACK_LIBRARY", self.spec["scalapack"].libs.joined(";")),
+                ]
+            )
```
Please correct me if im wrong but we probably need to define the blas and lapack prefixes for the serial variant.
EDIT: updated the libraries with the singular and plural names
```suggestion
        if "+mpi" in self.spec:
            # we use scalapack for linear algebra
            args.extend(
                [
                    self.define("SCALAPACK_FOUND", "true"),
                    self.define("SCALAPACK_INCLUDE_DIRS", self.spec["scalapack"].prefix.include),
                    self.define("SCALAPACK_LIBRARIES", self.spec["scalapack"].libs.joined(";")),
                    self.define("SCALAPACK_LIBRARY", self.spec["scalapack"].libs.joined(";")),
                ]
            )
        else:
            # we define the lapack and blas libraries
            lapack_libs = self.spec["lapack"].libs.joined(";")
            blas_libs = self.spec["blas"].libs.joined(";")
            args.extend(
                [
                    self.define("LAPACK_FOUND", True),
                    self.define("LAPACK_INCLUDE_DIRS", self.spec["lapack"].prefix.include),
                    self.define("LAPACK_LIBRARIES", lapack_libs),
                    self.define("LAPACK_LIBRARY", lapack_libs),
                    self.define("BLAS_FOUND", True),
                    self.define("BLAS_INCLUDE_DIRS", self.spec["blas"].prefix.include),
                    self.define("BLAS_LIBRARIES", blas_libs),
                    self.define("BLAS_LIBRARY", blas_libs),
                ]
            )
```

HaoZeke committed:
dftbp: Arpack for serial builds only

Co-authored-by: Ashwin Kumar Karnad <46030335+iamashwin99@users.noreply.github.com>

HaoZeke committed:
dftbp: Update option docs

Co-authored-by: Sebastian Ehlert <28669218+awvwgk@users.noreply.github.com>

HaoZeke committed:
dftbp: Update documentation for elsi

Co-authored-by: Ashwin Kumar Karnad <46030335+iamashwin99@users.noreply.github.com>

HaoZeke committed:
dftbp: Update the lapack and blas environment variables

Co-authored-by: Ashwin Kumar Karnad <46030335+iamashwin99@users.noreply.github.com>

HaoZeke committed:
dftbp: Add comment for context

Co-authored-by: Ashwin Kumar Karnad <46030335+iamashwin99@users.noreply.github.com>

HaoZeke committed:
dftbp: Tighter bounds on python

Co-authored-by: Ashwin Kumar Karnad <46030335+iamashwin99@users.noreply.github.com>

HaoZeke committed:
dftbp: Add negf only when shared

Co-authored-by: iamashwin99 <iamashwin99@users.noreply.github.com>

iamashwin99 commented:
LGTM

iamashwin99 commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,162 @@
+    variant(
+        "mbd",
+        default=False,
+        when="21.1:",
```
possibly typo
```suggestion
        when="@21.1:",
```

HaoZeke committed:
dftbp: Fix typo

Co-authored-by: Ashwin Kumar Karnad <46030335+iamashwin99@users.noreply.github.com>

aradi commented:
@iamashwin99 @HaoZeke @awvwgk : Thanks a lot for bringing this recipe forward!

iamashwin99 commented:
@HaoZeke  could you please check the checksums once, I am getting the following which is different for all the versions specified:
```shell
❯ spack checksum  dftbplus 22.1 21.2 21.1 20.2.1 20.2 20.1 19.1
==> Warning: Version 19.1 is deprecated
==> Selected 7 versions. 0 new versions
  22.1    https://github.com/dftbplus/dftbplus/releases/download/22.1/dftbplus-22.1.tar.xz
  21.2    https://github.com/dftbplus/dftbplus/releases/download/21.2/dftbplus-21.2.tar.xz
  21.1    https://github.com/dftbplus/dftbplus/releases/download/21.1/dftbplus-21.1.tar.xz
  20.2.1  https://github.com/dftbplus/dftbplus/releases/download/20.2.1/dftbplus-20.2.1.tar.xz
  20.2    https://github.com/dftbplus/dftbplus/releases/download/20.2/dftbplus-20.2.tar.xz
  20.1    https://github.com/dftbplus/dftbplus/releases/download/20.1/dftbplus-20.1.tar.xz
  19.1    https://github.com/dftbplus/dftbplus/releases/download/19.1/dftbplus-19.1.source.tar.xz  # NOTE: change of URL

==> Enter number of versions to take, or use a command:
    [c]hecksum  [e]dit  [f]ilter  [a]sk each  [n]ew only  [r]estart  [q]uit
action> c
==> Fetching https://github.com/dftbplus/dftbplus/releases/download/20.2/dftbplus-20.2.tar.xz
==> Fetching https://github.com/dftbplus/dftbplus/releases/download/21.1/dftbplus-21.1.tar.xz
==> Fetching https://github.com/dftbplus/dftbplus/releases/download/20.1/dftbplus-20.1.tar.xz
==> Fetching https://github.com/dftbplus/dftbplus/releases/download/22.1/dftbplus-22.1.tar.xz
==> Fetching https://github.com/dftbplus/dftbplus/releases/download/21.2/dftbplus-21.2.tar.xz
==> Fetching https://github.com/dftbplus/dftbplus/releases/download/20.2.1/dftbplus-20.2.1.tar.xz
==> Fetching https://github.com/dftbplus/dftbplus/releases/download/19.1/dftbplus-19.1.source.tar.xz

    version("22.1", sha256="02daca6f4c6372656598f3ba0311110c8e473c87c8d934d7bb276feaa4cc1c82")
    version("21.2", sha256="fbeb0e0ea93ab4dc4450f298ec712d2cf991f19f621badf57dae05f0e43b5906")
    version("21.1", sha256="8c1eb8a38f72c421e2ae20118a6db3a656fa84e8b180ef387e549a73ae77f970")
    version("20.2.1", sha256="95cc85fdb08bd57ca013bd09f4f902303720e17d015a5fab2d4db63fcb6d9cb3")
    version("20.2", sha256="eafd219159d600624041658046c89db539ceb0c1d2988b72321c80d9b992c9bf")
    version("20.1", sha256="04c2b906b8670937c8ddd9c5fb68e7e9921b464840cf54aa3d698db98167d0b7")
    version("19.1", sha256="78f45ef0571c78cf732a5493d32830455a832fa05ebcad43098895e46ad8d220")

```

HaoZeke committed:
dftbp: Update sha256

Co-authored-by: iamashwin99 <iamashwin99@users.noreply.github.com>

HaoZeke commented:
> @HaoZeke could you please check the checksums once, I am unable to compile with the current hashes, I am getting the following which is different for all the versions specified:
> 
> [BASH (SHELL) (30 lines)](#)
> ```shell
> ❯ spack checksum  dftbplus 22.1 21.2 21.1 20.2.1 20.2 20.1 19.1
> ==> Warning: Version 19.1 is deprecated
> ==> Selected 7 versions. 0 new versions
>   22.1    https://github.com/dftbplus/dftbplus/releases/download/22.1/dftbplus-22.1.tar.xz
>   21.2    https://github.com/dftbplus/dftbplus/releases/download/21.2/dftbplus-21.2.tar.xz
>   21.1    https://github.com/dftbplus/dftbplus/releases/download/21.1/dftbplus-21.1.tar.xz
>   20.2.1  https://github.com/dftbplus/dftbplus/releases/download/20.2.1/dftbplus-20.2.1.tar.xz
>   20.2    https://github.com/dftbplus/dftbplus/releases/download/20.2/dftbplus-20.2.tar.xz
>   20.1    https://github.com/dftbplus/dftbplus/releases/download/20.1/dftbplus-20.1.tar.xz
>   19.1    https://github.com/dftbplus/dftbplus/releases/download/19.1/dftbplus-19.1.source.tar.xz  # NOTE: change of URL
> 
> ==> Enter number of versions to take, or use a command:
>     [c]hecksum  [e]dit  [f]ilter  [a]sk each  [n]ew only  [r]estart  [q]uit
> action> c
> ==> Fetching https://github.com/dftbplus/dftbplus/releases/download/20.2/dftbplus-20.2.tar.xz
> ==> Fetching https://github.com/dftbplus/dftbplus/releases/download/21.1/dftbplus-21.1.tar.xz
> ==> Fetching https://github.com/dftbplus/dftbplus/releases/download/20.1/dftbplus-20.1.tar.xz
> ==> Fetching https://github.com/dftbplus/dftbplus/releases/download/22.1/dftbplus-22.1.tar.xz
> ==> Fetching https://github.com/dftbplus/dftbplus/releases/download/21.2/dftbplus-21.2.tar.xz
> ==> Fetching https://github.com/dftbplus/dftbplus/releases/download/20.2.1/dftbplus-20.2.1.tar.xz
> ==> Fetching https://github.com/dftbplus/dftbplus/releases/download/19.1/dftbplus-19.1.source.tar.xz
> 
>     version("22.1", sha256="02daca6f4c6372656598f3ba0311110c8e473c87c8d934d7bb276feaa4cc1c82")
>     version("21.2", sha256="fbeb0e0ea93ab4dc4450f298ec712d2cf991f19f621badf57dae05f0e43b5906")
>     version("21.1", sha256="8c1eb8a38f72c421e2ae20118a6db3a656fa84e8b180ef387e549a73ae77f970")
>     version("20.2.1", sha256="95cc85fdb08bd57ca013bd09f4f902303720e17d015a5fab2d4db63fcb6d9cb3")
>     version("20.2", sha256="eafd219159d600624041658046c89db539ceb0c1d2988b72321c80d9b992c9bf")
>     version("20.1", sha256="04c2b906b8670937c8ddd9c5fb68e7e9921b464840cf54aa3d698db98167d0b7")
>     version("19.1", sha256="78f45ef0571c78cf732a5493d32830455a832fa05ebcad43098895e46ad8d220")
> ```

Yup, thanks, fixed now.

HaoZeke commented:
@tldahlgren @awvwgk I think this is ready to be merged.

tldahlgren reviewed (commented):
I have (re)confirmed all of the sha256.

tldahlgren commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,162 @@
+
+    variant("shared", default=False, description="Most often for the Python wrappers.")
 
+    # ONLY for version 19.1
```
Should this comment reflect the when argument?

tldahlgren commented on var/spack/repos/builtin/packages/dftbplus/package.py:
```diff
@@ -3,85 +3,162 @@
-    depends_on("blas")
-    depends_on("scalapack", when="+mpi")
-    depends_on("mpi", when="+mpi")
+    depends_on("cmake@3.16:", type="build")
+    depends_on("ninja@1.10", type="build")
```
I think you want to consider using the ``when` context manager for dependencies that are specific to either build system.  There's an example in the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-dependencies.

iamashwin99 replied:
Ah yes, missed this
```suggestion
    depends_on("cmake@3.16:", type="build", when="@20.1:")
    depends_on("ninja@1.10", type="build", when="@20.1:")
```

tldahlgren reviewed (changes requested):
Minor comment change suggested but there *may* be some issues with the dependencies for one builder versus another.

I'm still trying to build the latest with these changes though my problems might related to my local configuration.

I still cannot build `19.1` though I supposed that could be a local issue for me as well:

```
spack install dftbplus@19.1+dftd3
==> Error: concretization failed for the following reasons:

   1. Cannot satisfy 'dftbplus@19.1'
```

HaoZeke committed:
dftpb: Add when directive for cmake and ninja

Co-authored-by: Ashwin Kumar Karnad <46030335+iamashwin99@users.noreply.github.com>

HaoZeke committed:
dftbp: Enforce comment

Co-authored-by: tldahlgren <tldahlgren@users.noreply.github.com>

iamashwin99 commented:
@tldahlgren does this look good for merging?
